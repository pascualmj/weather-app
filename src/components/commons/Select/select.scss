@import "../../../assets/scss/variables";

.select {
  height: 5rem;
  border: 0.2rem solid $color-full;
  position: relative;
  cursor: pointer;

  .option-selected {
    height: 4.6rem;
    padding: 0 1.5rem;
    display: flex;
    justify-content: space-between;
    align-items: center;

    p {
      flex: 1 1 auto;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }

    .material-icons {
      flex: 0 0 auto;
      font-size: $text-subtitle;
      margin-left: 2rem;
    }
  }

  ul.options {
    height: auto;
    max-height: 27.8rem;
    overflow-y: auto;
    position: absolute;
    left: -0.2rem;
    right: -0.2rem;
    top: 4.6rem;
    border: 2px solid $color-full;
    border-top: none;
    display: none;

    &.active {
      display: block;
    }

    li {
      height: 4.6rem;
      padding: 0 1.5rem;
      display: flex;
      align-items: center;
      background-color: $color-lighter;
      color: $color-secondary;

      &:hover {
        color: $color-full;
      }

      &.selected {
        background-color: $color-absence;
        color: $color-full;
      }
    }
  }
}
