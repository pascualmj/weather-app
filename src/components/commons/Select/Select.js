import React, { useState, useRef } from "react";
import PropTypes from "prop-types";
import "./select.scss";

import Text from "../Text";

import { useOutsideListener } from "../../../hooks";

const Select = ({ options, value, onChange }) => {
  const [isOpen, setIsOpen] = useState(false);
  const wrapperElement = useRef(null);
  useOutsideListener(wrapperElement, handleClose);

  function handleClose() {
    setIsOpen(false);
  }

  const handleToggle = () => {
    setIsOpen(!isOpen);
  };

  const handleClick = (value) => {
    return (e) => {
      onChange(value);
      setIsOpen(false);
    };
  };

  return (
    <div className="select" ref={wrapperElement}>
      <div className="option-selected" onClick={handleToggle}>
        <Text>{options.find((opt) => opt.value === value).label}</Text>
        <span className="material-icons">unfold_more</span>
      </div>
      <ul className={`options ${isOpen ? "active" : ""}`}>
        {options.map((opt) => (
          <li
            key={opt.value}
            className={opt.value === value ? "selected" : ""}
            onClick={handleClick(opt.value)}
          >
            {opt.label}
          </li>
        ))}
      </ul>
    </div>
  );
};

Select.propTypes = {
  options: PropTypes.array.isRequired,
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
};

export default Select;
