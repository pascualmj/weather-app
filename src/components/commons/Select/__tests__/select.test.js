import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import Select from "../Select";

describe("Component: Select", () => {
  const setupProps = () => ({
    options: [
      {
        value: "a",
        label: "a",
      },
      {
        value: "b",
        label: "b",
      },
    ],
    value: "a",
    onChange: jest.fn(),
  });

  test("Should match snapshot.", () => {
    const { asFragment } = render(<Select {...setupProps()} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
