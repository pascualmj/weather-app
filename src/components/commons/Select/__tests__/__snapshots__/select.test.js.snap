// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Select Should match snapshot. 1`] = `
<DocumentFragment>
  <div
    class="select "
  >
    <div
      class="option-selected"
    >
      <p
        class="text color-full size-regular weight-regular "
      >
        a
      </p>
      <span
        class="material-icons"
      >
        unfold_more
      </span>
    </div>
    <ul
      class="options "
    >
      <li
        class="selected"
      >
        a
      </li>
      <li
        class=""
      >
        b
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
